ramips: add support for ipTIME A3002MESH

koreapyj opened the pull request:
Add support for ipTIME A3002MESH.

Hardware:
- SoC: MediaTek MT7621AT (880MHz, Duel-Core)
- RAM: DDR3 128MB
- Flash: XMC XM25QH128AHIG (SPI-NOR 16MB)
- WiFi: MediaTek MT7615D (2.4GHz, 5GHz, DBDC)
- Ethernet: MediaTek MT7530 (WAN x1, LAN x2, SoC)
- UART: [GND, RX, TX, 3.3V] (57600 8N1, J4)

MAC addresses:
| interface |        MAC        |     source     | comment
|-----------|-------------------|----------------|----------
|       LAN | 70:XX:XX:5X:XX:X3 |                |
|       WAN | 70:XX:XX:5X:XX:X1 | u-boot 0x1fc40 |
|   WLAN 2G | 72:XX:XX:4X:XX:X0 |                |
|   WLAN 5G | 70:XX:XX:5X:XX:X0 | factory 0x4    |
|           | 70:XX:XX:5X:XX:X0 | u-boot 0x1fc20 | unknown
|           | 70:XX:XX:5X:XX:X2 | factory 0x8004 | unknown

- WLAN 2G MAC address is not the same as stock firmware since OpenWrt uses LAN MAC address with local bit sets.

Installation:
1. Flash initramfs image. This can be done using stock web ui or TFTP
2. Connect to OpenWrt with an SSH connection to 192.168.1.1
3. Perform sysupgrade with sysupgrade image

Revert to stock firmware:
- Flash stock firmware via OEM TFTP Recovery mode
- Perform sysupgrade with stock image

TFTP Recovery method:
1. Unplug the router
2. Hold the reset button and plug in
3. Release when the power LED stops flashing and go off
4. Set your computer IP address manually to 192.168.0.x / 255.255.255.0
5. Flash image with TFTP client to 192.168.0.1

musashino205 commented on target/linux/ramips/dts/mt7621_iptime_a3002me.dts:
```diff
@@ -0,0 +1,167 @@
+				reg = <0x0 0x20000>;
+				read-only;
+
+                compatible = "nvmem-cells";
+                #address-cells = <1>;
+                #size-cells = <1>;
+
+                macaddr_uboot_1fc20: macaddr@1fc20 {
+                    reg = <0x1fc20 0x6>;
+                };
+
+                macaddr_uboot_1fc40: macaddr@1fc40 {
+                    reg = <0x1fc40 0x6>;
+                };
```
wrong indentation

musashino205 replied:
And these two MAC addresses are not used?

koreapyj replied:
It seems like stock firmware use them but in my dts they are not used. removed.

musashino205 commented on target/linux/ramips/dts/mt7621_iptime_a3002me.dts:
```diff
@@ -0,0 +1,167 @@
+				reg = <0x30000 0x10000>;
+				read-only;
+
+                compatible = "nvmem-cells";
+                #address-cells = <1>;
+                #size-cells = <1>;
+
+                macaddr_factory_4: macaddr@4 {
+                    reg = <0x4 0x6>;
+                };
+
+                macaddr_factory_8004: macaddr@8004 {
+                    reg = <0x8004 0x6>;
+                };
```
wrong indentation

mans0n commented on target/linux/ramips/dts/mt7621_iptime_a3002me.dts:
```diff
@@ -0,0 +1,151 @@
+#include <dt-bindings/leds/common.h>
+
+/ {
+	compatible = "iptime,a3002me", "mediatek,mt7621-soc";
```
No, this should be `iptime,a3002mesh`.

koreapyj replied:
But stock uImage introduce self as "a3002me". I'm little bit confused.

![image](https://user-images.githubusercontent.com/1674267/153390259-eddd3b86-2112-4e11-988b-25f921660824.png)

koreapyj replied:
https://download.iptime.co.kr/online_upgrade/a3002me_ml_14_124.bin

mans0n replied:
This string is for OpenWrt, not iptime code name. You should keep the `UIMAGE_NAME`, of course.

mans0n commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -757,6 +757,16 @@ define Device/iodata_wnpr2600g
 endef
 TARGET_DEVICES += iodata_wnpr2600g
 
+define Device/iptime_a3002me
```
Ditto.

mans0n commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -39,6 +39,7 @@ ramips_setup_interfaces()
 		;;
 	asiarf,ap7621-nv1|\
 	glinet,gl-mt1300|\
+	iptime,a3002me|\
```
Ditto.

mans0n reviewed (commented):
Hi, mostly seems good. A few comments below.

mans0n commented:
> Revert to stock firmware:
>     1. Flash stock firmware via OEM's Recovery mode
>     2. Perform sysupgrade with stock image

I guess user should take either one of the two paths, so numbered list is inappropriate here?

And I'm pretty sure this DBDC also has an unpredictable MAC address scheme. Could you list its MAC addresses in the commit message?

koreapyj commented:
> > Revert to stock firmware:
> > 1. Flash stock firmware via OEM's Recovery mode
> > 2. Perform sysupgrade with stock image
> 
> I guess user should take either one of the two paths, so numbered list might be inappropriate here?
> 
> And I'm pretty sure this DBDC also has an unpredictable MAC address scheme. Could you list its MAC addresses in the commit message?

fixed.

mans0n commented:
> `|       WAN | 70:XX:XX:XX:XX:X1 | u-boot 0x1fc40 |`

`mac-address-increment` could do the job, but I would prefer to use the exact match if possible.
ipTIME's convention was to use 0x1fc20 for LAN and 0x1fc40 for WAN, but the scheme has changed since the introduction of DBDC.

> - WLAN 2G MAC address may not be the same as stock firmware due to DBDC.

b8168f4716ad used LAN address with local bit set for wlan2g address, so let's follow the path. Take a look how I dealt with it in #4915.

koreapyj commented:
> mac-address-increment could do the job, but I would prefer to use the exact match if possible.
I also prefer that but config is little bit weird on this device. U-boot always complain about CRC mismatch (even if device is in untouched state) and use default config. So it may be a trouble if we use MAC address from config partition.

mans0n commented:
> So it may be a trouble if we use MAC address from config partition.

Why use config partiton? The WAN address should be in the u-boot partition, isn't it?

koreapyj commented:
> Why use config partiton? The WAN address should be in the u-boot partition, isn't it?

You're right. I confused with other board. I'll fix it.

mans0n commented:
> But I can't reproduce PHY1 MAC address from stock firmware with your commit. Should I have to leave it different?

Do you mean wlan2g or wlan5g?
wlan2g will have a different MAC address from stock FW, but wlan5g should be same.

koreapyj commented:
> > But I can't reproduce PHY1 MAC address from stock firmware with your commit. Should I have to leave it different?
>
> Do you mean wlan2g or wlan5g?
> wlan2g will have a different MAC address from stock FW, but wlan5g should be same.

I mean wlan2g. Your commit produces exactly same result as current codes.

I don't know why but I think current `/etc/hotplug.d/ieee80211/10_fix_wifi_mac` is not working correctly. It should swap MAC address of phy0 and phy1 but no differences at all. Did I miss something?

koreapyj commented:
> Err, I thought wlan2g/wlan5g corresponds to phy0/phy1 for mt7615 DBDC, am I wrong?
> IIRC `10_fix_wifi_mac` is applied only when the wifi is actually turned on, that might be the reason?

`10_fix_wifi_mac` works correctly but `iwinfo` shows original mac address so I confused.

Anyway, `macaddr_setbit_la "$(macaddr_unsetbit $hw_mac_addr 28)" > /sys${DEVPATH}/macaddress` is produces same result as stock firmware on my board but I concern that it is just "bitwise not" operation. If board has even value on first number of 4th bytes it will not working correctly.

Then `macaddr_setbit_la "$(macaddr_add $hw_mac_addr -1048576)" > /sys${DEVPATH}/macaddress` should work but if board has value lower than 0 it will screw everything.

I have no idea how other boards' macs looks like but I think both risky.

koreapyj commented:
If difference is acceptable I'll push new commit.

mans0n commented:
> `10_fix_wifi_mac` works correctly but `iwinfo` shows original mac address so I confused.

It shouldn't... I always double-check `ip link` with `iwinfo`.

> Anyway, `macaddr_setbit_la "$(macaddr_unsetbit $hw_mac_addr 28)" > /sys${DEVPATH}/macaddress` is produces same result as stock firmware on my board but I concern that it is just "bitwise and not" operation. If board has even value on first number of 4th bytes it will not working correctly.
> 
> Then `macaddr_setbit_la "$(macaddr_add $hw_mac_addr -1048576)" > /sys${DEVPATH}/macaddress` should work but if board has value lower than 0 it will screw everything.
> 
> I have no idea how other boards' macs looks like but I think both risky.

That's exactly why we chose to use LAN address with local bit set instead of trying to reproduce the exact address for wlan2g in b8168f4716ad. I thought that will lead to the same config in #4915, no?

koreapyj commented:
Okay, all issues now fixed. Still `iwinfo` reports wrong MAC address when no ap configured but I think it cannot be handled with `10_fix_wifi_mac`, right?

koreapyj committed:
ramips: add support for ipTIME A3002MESH
Add support for ipTIME A3002MESH.

Hardware:
- SoC: MediaTek MT7621AT (880MHz, Duel-Core)
- RAM: DDR3 128MB
- Flash: XMC XM25QH128AHIG (SPI-NOR 16MB)
- WiFi: MediaTek MT7615D (2.4GHz, 5GHz, DBDC)
- Ethernet: MediaTek MT7530 (WAN x1, LAN x2, SoC)
- UART: [GND, RX, TX, 3.3V] (57600 8N1, J4)

MAC addresses:
| interface |        MAC        |     source     | comment
|-----------|-------------------|----------------|----------
|       LAN | 70:XX:XX:5X:XX:X3 |                |
|       WAN | 70:XX:XX:5X:XX:X1 | u-boot 0x1fc40 |
|   WLAN 2G | 72:XX:XX:4X:XX:X0 |                |
|   WLAN 5G | 70:XX:XX:5X:XX:X0 | factory 0x4    |
|           | 70:XX:XX:5X:XX:X0 | u-boot 0x1fc20 | unknown
|           | 70:XX:XX:5X:XX:X2 | factory 0x8004 | unknown

- WLAN 2G MAC address is not the same as stock firmware since OpenWrt uses LAN MAC address with local bit sets.

Installation:
1. Flash initramfs image. This can be done using stock web ui or TFTP
2. Connect to OpenWrt with an SSH connection to 192.168.1.1
3. Perform sysupgrade with sysupgrade image

Revert to stock firmware:
- Flash stock firmware via OEM TFTP Recovery mode
- Perform sysupgrade with stock image

TFTP Recovery method:
1. Unplug the router
2. Hold the reset button and plug in
3. Release when the power LED stops flashing and go off
4. Set your computer IP address manually to 192.168.0.x / 255.255.255.0
5. Flash image with TFTP client to 192.168.0.1

Signed-off-by: Yoonji Park <koreapyj@dcmys.kr>

mans0n commented:
Thanks! Merged your patch as 125b9aec2929.
